feat: circular autofix/conflicting rules detection (#19514)
eslint/eslint: docs/src/use/troubleshooting/circular-fixes.md

---
title: Circular fixes detected …
eleventyNavigation:
    key: circular fixes
    parent: troubleshooting
    title: Circular fixes detected …
---

## Symptoms

When running ESLint with the `--fix` option, you may see the following warning:

```plaintext
ESLintCircularFixesWarning: Circular fixes detected while fixing path/to/file. It is likely that you have conflicting rules in your configuration.
```

## Cause

> You have conflicting fixable rules in your configuration. ESLint autofixes code in mulitple passes. In one pass, one of the rules changes the code, and in the following pass, another rule changes the new code back to the previous version. ESLint detects that autofix is going in circles and emits this warning.

Reviewer: Just to clarify a bit because this is an abstract concept.

```suggestion
You have conflicting fixable rules in your configuration. ESLint autofixes code in multiple passes, meaning it's possible that a fix in one pass is undone in a subsequent pass. For example, in the first pass a rules removes a trailing comma and in the following pass a different rule adds a trailing comma in the same place, effectively changing the code back to the previous version. ESLint emits a warning when it detects cycles like this.
```

> You have conflicting fixable rules in your configuration. ESLint autofixes code in multiple passes, meaning it's possible that a fix in one pass is undone in a subsequent pass. For example, in the first pass a rules removes a trailing comma and in the following pass a different rule adds a trailing comma in the same place, effectively changing the code back to the previous version. ESLint emits a warning when it detects cycles like this.

Author: ```suggestion
You have conflicting fixable rules in your configuration. ESLint autofixes code in multiple passes, meaning it's possible that a fix in one pass is undone in a subsequent pass. For example, in the first pass a rule removes a trailing comma and in the following pass a different rule adds a trailing comma in the same place, effectively changing the code back to the previous version. ESLint emits a warning when it detects cycles like this.
```

Typo.

## Resolution

Common resolutions for this issue include:

* Remove or reconfigure one of the conflicting rules in your configuration file.

How to find the conflicting rules:

1. Open the file specified in the warning in an editor that supports applying individual fixes (for example, VS Code).
1. In the list of lint problems, find a fixable rule. That is one of the conflicting rules.

> 1. Apply the fix ("Fix this rule-name problem" action in VS Code).

Reviewer: For clarity

```suggestion
1. Apply the fix ("Fix this (rule-name) problem" action in VS Code).
```
